### ISSUE TITLE
simplify?

### ISSUE THREAD

**Issue by rchl** [predragnikolic/JS-Template-String-Converter#3]
- is `set_timeout` needed?
- should it return after triggering a command instead of processing further characters?
- for converting TO a template literal, would it be enough to handle that from the keybinding itself or would that miss some cases? That change listener is fairly simply but if I would use such a plugin, I would probably still prefer not having it and I would be fine with no conversion to regular string on removing the last occurrence of `${}`

https://github.com/predragnikolic/JS-Template-String-Converter/blob/da6f172f4f5bf037d9b2e4679aa75a4138ec969e/plugin.py#L14-L19

**Comment by predragnikolic** [comment]
All good points, 
when I grab some time will look into it

**Comment by predragnikolic** [comment]
I managed to simplify it, by using the `chain` command
<details>
<summary>plugin.py</summary>

```py
from typing import Optional
import sublime
import sublime_plugin


class ConvertToTemplateString(sublime_plugin.TextCommand):
    """ This command is guaranteed to executed when { is pressed """
    def run(self, edit):
        point = get_cursor_point(self.view)
        if not point:
            return
        if not in_supported_file(self.view, point):
            return None
        regular_string_region = get_regular_string_region(self.view, point)
        if not regular_string_region:
            return
        scan_region = self.view.substr(sublime.Region(point-2, point+2))
        # the user could typed $ or {
        if "${" not in scan_region:
            return
        first_quote = regular_string_region.begin()
        last_quote = regular_string_region.end() - 1
        are_quotes = is_regular_quote(self.view.substr(first_quote)) and is_regular_quote(self.view.substr(last_quote))
        if not are_quotes:
            return # sanity check, just to 100% make sure we are replacing quotes
        if is_jsx_attribute(self.view, point) and not is_jsx_attribute_wrapped_with_curly_brackets(self.view, point):
            # insert surrounding curly brackets
            self.view.replace(
                edit, sublime.Region(last_quote, last_quote + 1), '`}')
            self.view.replace(
                edit, sublime.Region(first_quote, first_quote + 1), '{`')
            return
        self.view.replace(
            edit, sublime.Region(last_quote, last_quote + 1), '`')
        self.view.replace(
            edit, sublime.Region(first_quote, first_quote + 1), '`')


def is_jsx_attribute(view: sublime.View, point: int) -> bool:
    return view.match_selector(point, "meta.jsx meta.tag.attributes")


def is_jsx_attribute_wrapped_with_curly_brackets(view: sublime.View, point: int) -> bool:
    return view.match_selector(point, "meta.jsx meta.tag.attributes meta.interpolation meta.string string.quoted")


def get_cursor_point(view: sublime.View) -> Optional[int]:
    sel = view.sel()
    if not sel:
        return
    return sel[0].b


def is_regular_quote(char: str) -> bool:
    return char in ["'", '"']


def in_supported_file(view: sublime.View, point: int) -> bool:
    return view.match_selector(point, "source.js | source.jsx | source.ts | source.tsx | text.html.ngx | text.html.svelte | text.html.vue")


def get_regular_string_region(view: sublime.View, point: int) -> Optional[sublime.Region]:
    return view.expand_to_scope(point, "string.quoted.single | string.quoted.double")
```

keybindings:

```json
[
    {"keys": ["{"], "command": "chain",
        "args": {
            "commands": [
                {"command": "insert_snippet", "args": {"contents": "{$0}"}},
                {"command": "convert_to_template_string"},
            ]
        },
        "context": [
            { "key": "setting.auto_match_enabled" },
            { "key": "selector", "operand": "source.js | source.jsx | source.ts | source.tsx | text.html.ngx | text.html.svelte | text.html.vue" },
            { "key": "selection_empty", "match_all": true },
            { "key": "preceding_text", "operator": "not_regex_contains", "operand": "\\\\$", "match_all": true },
    ]}
]
```

i can go one step further and remove the imports from `typing`

</details>

**Comment by predragnikolic** [comment]
closed with [bae810d](https://github.com/predragnikolic/JS-Template-String-Converter/commit/bae810d50a187b9dc170e10c6eb9a79f37c72014)
